services/typescript-service.ts

```diff
@@ -0,0 +1,316 @@
+import * as path from "path";
+import * as os from "os";
+import temp = require("temp");
+import {exportedPromise} from "../decorators";
+temp.track();
+
+interface ITypeScriptCompilerMessages {
+	level1ErrorCount: number;
+	level5ErrorCount: number;
+	nonEmitPreventingWarningCount: number;
+	hasPreventEmitErrors: boolean;
+}
+
+interface ITypeScriptCompilerSettings {
+	pathToCompiler: string;
+	version: string;
+}
+
+export class TypeScriptService implements ITypeScriptService {
+	private static DEFAULT_TSC_VERSION = "1.8.10";
+	private static TYPESCRIPT_MODULE_NAME = "typescript";
+
+	private typeScriptFiles: string[];
+	private definitionFiles: string[];
+	private noEmitOnError: boolean;
+	private hasInstalledTsc: boolean;
+
+	constructor(private $childProcess: IChildProcess,
+		private $fs: IFileSystem,
+		private $logger: ILogger,
+		private $npmService: INpmService) {
+		this.hasInstalledTsc = false;
+	}
+
+	@exportedPromise("typeScriptService")
+	public transpile(projectDir: string, typeScriptFiles?: string[], definitionFiles?: string[], options?: ITypeScriptTranspileOptions): IFuture<string> {
+		return ((): string => {
+			options = options || {};
+			let compilerOptions = this.getCompilerOptions(projectDir, options).wait();
+			this.noEmitOnError = compilerOptions.noEmitOnError;
+			this.typeScriptFiles = typeScriptFiles || [];
+			this.definitionFiles = definitionFiles || [];
+			if (this.typeScriptFiles.length > 0) {
+				// Create typeScript command file
+				let typeScriptCommandsFilePath = path.join(temp.mkdirSync("typeScript-compilation"), "tscommand.txt");
+				let typeScriptCompilerOptions = this.getTypeScriptCompilerOptionsAsArguments(compilerOptions);
+				let typeScriptDefinitionsFiles: string[];
+
+				if (!this.hasTsConfigFile(projectDir).wait()) {
+					typeScriptDefinitionsFiles = this.getDefaultTypeScriptDefinitionsFiles(options.pathToDefaultDefinitionFiles).wait();
+				}
+
+				this.$fs.writeFile(typeScriptCommandsFilePath, this.typeScriptFiles.concat(typeScriptDefinitionsFiles).concat(typeScriptCompilerOptions).join(" ")).wait();
+
+				let typeScriptCompilerSettings = this.getTypeScriptCompiler().wait();
+
+				// Log some messages
+				this.$logger.out("Compiling...".yellow);
+				_.each(this.typeScriptFiles, file => {
+					this.$logger.out(`### Compile ${file}`.cyan);
+				});
+				this.$logger.out(`Using tsc version ${typeScriptCompilerSettings.version}`.cyan);
+
+				// Core compilation
+				return this.runTranspilation(projectDir, typeScriptCompilerSettings.pathToCompiler, { typeScriptCommandsFilePath }).wait();
+			} else {
+				return this.transpileWithDefaultOptions(projectDir, compilerOptions).wait();
+			}
+		}).future<string>()();
+	}
+
+	// Uses tsconfig.json if it exists
+	public transpileWithDefaultOptions(projectDir: string, options?: ITypeScriptTranspileOptions): IFuture<string> {
```

Comment:
I could be mistaken, but I believe you can safely merge this method's implementation into the more generic `transpile` method and delete this